subsys/dfu/boot/mcuboot.c

```diff
@@ -162,36 +165,89 @@ static int boot_magic_write(u32_t bank_offs)
 	return rc;
 }
 
-static int boot_magic_code_check(const u32_t *magic)
+int boot_read_bank_header(u32_t bank_offset,
+			  struct mcuboot_img_header *header,
+			  size_t header_size)
 {
-	int i;
-
-	if (memcmp(magic, boot_img_magic, sizeof(boot_img_magic)) == 0) {
-		return BOOT_MAGIC_GOOD;
+	int rc;
+	u32_t magic;
+	/*
+	 * Raw representation of the image header.
+	 *
+	 * The header_magic and header_size fields are only needed for
+	 * sanity checking that bank_offset points at a supported
+	 * image header.
+	 *
+	 * These are unsupported by MCUboot or not relevant to Zephyr now:
+	 *
+	 * - image_load_address only matters for PIC code (unsupported)
+	 * - image_flags are unsupported / currently irrelevant to Zephyr
+	 *
+	 * These are potentially useful:
+	 *
+	 * - image_size
+	 * - version
+	 */
+	struct {
+		u32_t header_magic;
+		u32_t image_load_address;
+		u16_t header_size;
+		u16_t pad;
+		u32_t image_size;
+		u32_t image_flags;
+		struct {
+			u8_t major;
+			u8_t minor;
+			u16_t revision;
+			u32_t build_num;
+		} version;
+		u32_t pad2;
+	} v1_raw;
```

Comment:
This needs to be __packed

Reply:
And that probably means that you will need to use the `sys_get_*` macros in [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/misc/byteorder.h) to access the members

Reply:
No objections to adding __packed, but I don't see why it "needs" this attribute. The on-flash layout is chosen so the struct is already packed. What am I missing?

Reply:
I.e. fields fall at offsets that are all aligned with their sizes

Reply:
The C specification allows the compiler to pad freely unless you specify packed, no matter what offsets or sizes are present. Unless you specify packed a compiler could add padding anywhere, so specifying packed is required always when mapping a struct to serialized data, no matter any other consideration

Reply:
Also you need to take into account endianness as well, this will not work on both endiannesses as is

Reply:
> The C specification allows the compiler to pad freely unless you specify packed, no matter what offsets or sizes are present.

You're right! I always thought that the compiler was forced  to pack along alignment boundaries if possible, but reading the standard closely, it's aligned + implementation-defined, with no text forbidding holes even if the fields could lie contiguously. Huh.



> Also you need to take into account endianness as well, this will not work on both endiannesses as is

Yeah, that's my bad reading the mcuboot design doc. The magic number written in the trailer is host byte order, and for some reason I thought the rest of these fields were too. The header is little endian though, so this does need byte order conversion.